### PR TITLE
[2.3.2.r1.4] BACKPORT: support adoptable storage on FBE

### PR DESCRIPTION
**Pull request by lzlrd**
This complements the patches I made sonyxperiadev/device-sony-yoshino#133 and sonyxperiadev/device-sony-common#597.

A clean re-implementation I did and fixed builds.

Tested builds:
**Kirin (Ganges)**
```
  GEN     .version
  LD      vmlinux.o
  MODPOST vmlinux.o
  CHK     include/generated/compile.h
  UPD     include/generated/compile.h
  CC      init/version.o
  LD      init/built-in.o
  KSYM    .tmp_kallsyms1.o
  KSYM    .tmp_kallsyms2.o
  LD      vmlinux
  SORTEX  vmlinux
  SYSMAP  System.map
  OBJCOPY arch/arm64/boot/Image
  DTC     arch/arm64/boot/dts/qcom/sdm630-ganges-kirin-sde.dtb
  Building modules, stage 2.
  MODPOST 0 modules
  GZIP    arch/arm64/boot/Image.gz
  CAT     arch/arm64/boot/Image.gz-dtb
make[1]: Leaving directory '/home/lazerl0rd/recovery/kernel/sony/msm/out'
lazerl0rd@build > ~/android/recovery/kernel/sony/msm
$ date
Tue Apr 23 15:47:11 CDT 2019
```
**Maple (Yoshino)**
```
  GEN     .version
  LD      vmlinux.o
  MODPOST vmlinux.o
  CHK     include/generated/compile.h
  UPD     include/generated/compile.h
  CC      init/version.o
  LD      init/built-in.o
  KSYM    .tmp_kallsyms1.o
  KSYM    .tmp_kallsyms2.o
  LD      vmlinux
  SORTEX  vmlinux
  SYSMAP  System.map
  OBJCOPY arch/arm64/boot/Image
  DTC     arch/arm64/boot/dts/qcom/msm8998-v2.1-yoshino-maple-sde_generic.dtb
  DTC     arch/arm64/boot/dts/qcom/msm8998-v2-yoshino-maple-sde_generic.dtb
  Building modules, stage 2.
  MODPOST 0 modules
  GZIP    arch/arm64/boot/Image.gz
  CAT     arch/arm64/boot/Image.gz-dtb
make[1]: Leaving directory '/home/lazerl0rd/recovery/kernel/sony/msm/out'
lazerl0rd@build > ~/android/recovery/kernel/sony/msm
$ date
Tue Apr 23 15:37:02 CDT 2019
```
_Toolchain used: AOSP GCC 4.9 (reverted deprecation patch)._

I cannot currently test "on-device" per se, but am able to provide built images for those who want to test. Apologies again for yesterday's SODP-killer.

### CODE REVIEW

File: fs/ext4/inode.c
Comment (nathanchance) [inline fs/ext4/inode.c:3833]: This is still wrong compared to the original patch: https://android-review.googlesource.com/c/kernel/common/+/593373

There should only be one `ll_rw_block` call.

File: fs/ext4/inode.c
Comment (nathanchance) [inline fs/ext4/inode.c]: I don't see much of a reason to have parentheses around `decrypt`, especially since `&& !fscrypt_using_hardware_encryption(inode)` should probably be added to `decrypt`'s initialization like I mention above.

File: fs/ext4/inode.c
Comment (nathanchance) [inline fs/ext4/inode.c:3830]: Based on the value of `decrypt` in `ext4_block_write_begin`, this should probably be the same:

```
			decrypt = S_ISREG(inode->i_mode) &&
				ext4_encrypted_inode(inode) &&
				!fscrypt_using_hardware_encryption(inode);
```

**Review by nathanchance — COMMENTED** [review]
Where is the `fs/crypto/bio.c` change from https://android-review.googlesource.com/c/kernel/common/+/593373?

It's not usually wise to omit a change to a file (especially core files like `fs/`) without a justification, which I don't see in your commit message or pull message.

I can't obviously comment on the correctness of the patches, just how they were backported.

File: fs/ext4/inode.c
Comment (lzlrd) [reply fs/ext4/inode.c:3830]: It's different in the original commit though, I'm uncertain of changing it from that. There must've been some reason it's different, no? Especially since ext4 maintainer took a look at this.

File: fs/ext4/inode.c
Comment (lzlrd) [reply fs/ext4/inode.c]: It didn't build without the parentheses.

File: fs/ext4/inode.c
Comment (lzlrd) [reply fs/ext4/inode.c:3833]: Yupp, this is a mistake on my side. I'll force-push a fix now.

Comment (lzlrd) [reply fs/ext4/inode.c:3833]: Fixed it @nathanchance.

File: fs/ext4/inode.c
Comment (nathanchance) [reply fs/ext4/inode.c:3830]: It's because of CAF's additions, c692cb96096def8727ae342ea0926c651f2eeed7 and 36c6512ecdc28e18befec67802af4ed075772112.

Comment (lzlrd) [reply fs/ext4/inode.c:3830]: > It's because of CAF's additions, [c692cb9](https://github.com/sonyxperiadev/kernel/commit/c692cb96096def8727ae342ea0926c651f2eeed7) and [36c6512](https://github.com/sonyxperiadev/kernel/commit/36c6512ecdc28e18befec67802af4ed075772112).

Ah, okai. The ICE commits. I see now. I'm going to do this as a seperate commit to the others as it diverges from the original.

File: fs/ext4/inode.c
Comment (nathanchance) [reply fs/ext4/inode.c]: Are you sure?

This change compiles fine for me with `aosp_yoshino_maple_defconfig`:

```diff
diff --git a/fs/ext4/inode.c b/fs/ext4/inode.c
index 7cf1a56c93be..e7f054639d28 100644
--- a/fs/ext4/inode.c
+++ b/fs/ext4/inode.c
@@ -3835,7 +3835,7 @@ static int __ext4_block_zero_page_range(handle_t *handle,
                /* Uhhuh. Read error. Complain and punt. */
                if (!buffer_uptodate(bh))
                        goto unlock;
-               if ((decrypt) &&
+               if (decrypt &&
                    !fscrypt_using_hardware_encryption(inode)) {
                        /* We expect the key to be set. */
                        BUG_ON(!fscrypt_has_encryption_key(inode));

```

Comment (lzlrd) [reply fs/ext4/inode.c]: Oh, that's what you meant. Sorry I really misunderstood you there.

File: fs/ext4/inode.c
Comment (lzlrd) [reply fs/ext4/inode.c:3830]: Done

File: fs/ext4/inode.c
Comment (lzlrd) [reply fs/ext4/inode.c]: Well, I now used your initialization method, so I'd say this is done.

**Comment by lzlrd** [comment]
As of https://github.com/sonyxperiadev/kernel/commit/1aa83e5614ca429c9aac0535de6e53db377efe36, we have a build working out:

```
  GEN     .version
  LD      vmlinux.o
  MODPOST vmlinux.o
  CHK     include/generated/compile.h
  UPD     include/generated/compile.h
  CC      init/version.o
  LD      init/built-in.o
  KSYM    .tmp_kallsyms1.o
  KSYM    .tmp_kallsyms2.o
  LD      vmlinux
  SORTEX  vmlinux
  SYSMAP  System.map
  OBJCOPY arch/arm64/boot/Image
  DTC     arch/arm64/boot/dts/qcom/msm8998-v2.1-yoshino-maple-sde_generic.dtb
  DTC     arch/arm64/boot/dts/qcom/msm8998-v2-yoshino-maple-sde_generic.dtb
  Building modules, stage 2.
  MODPOST 0 modules
  GZIP    arch/arm64/boot/Image.gz
  CAT     arch/arm64/boot/Image.gz-dtb
make[1]: Leaving directory '/home/lazerl0rd/recovery/kernel/sony/msm/out'
lazerl0rd@build > ~/android/recovery/kernel/sony/msm
$ date
Tue Apr 23 17:27:16 CDT 2019
```

**Comment by lzlrd** [comment]
> Where is the `fs/crypto/bio.c` change from https://android-review.googlesource.com/c/kernel/common/+/593373?
> 
> It's not usually wise to omit a change to a file (especially core files like `fs/`) without a justification, which I don't see in your commit message or pull message.
> 
> I can't obviously comment on the correctness of the patches, just how they were backported.

Ah, my apologies, I seem to have missed out this comment. Well it's 100% due to changing `bio_set_op_attrs(bio, REQ_OP_WRITE, 0);` to `bio_set_op_attrs(bio, REQ_OP_WRITE, REQ_NOENCRYPT);` causing this (below):

```
In file included from ../include/asm-generic/bug.h:4:0,
                 from ../arch/arm64/include/asm/bug.h:67,
                 from ../include/linux/bug.h:4,
                 from ../include/linux/mmdebug.h:4,
                 from ../include/linux/mm.h:8,
                 from ../include/linux/pagemap.h:7,
                 from ../fs/crypto/bio.c:22:
../fs/crypto/bio.c: In function 'fscrypt_zeroout_range':
../include/linux/compiler.h:530:38: error: call to '__compiletime_assert_134' declared with attribute error: BUILD_BUG_ON failed: ((1ULL << __REQ_NOENCRYPT)) + 0U >= (1U << BIO_OP_SHIFT)
  _compiletime_assert(condition, msg, __compiletime_assert_, __LINE__)
                                      ^
../include/linux/compiler.h:513:4: note: in definition of macro '__compiletime_assert'
    prefix ## suffix();    \
    ^
../include/linux/compiler.h:530:2: note: in expansion of macro '_compiletime_assert'
  _compiletime_assert(condition, msg, __compiletime_assert_, __LINE__)
  ^
../include/linux/bug.h:54:37: note: in expansion of macro 'compiletime_assert'
 #define BUILD_BUG_ON_MSG(cond, msg) compiletime_assert(!(cond), msg)
                                     ^
../include/linux/bug.h:78:2: note: in expansion of macro 'BUILD_BUG_ON_MSG'
  BUILD_BUG_ON_MSG(condition, "BUILD_BUG_ON failed: " #condition)
  ^
../include/linux/blk_types.h:117:3: note: in expansion of macro 'BUILD_BUG_ON'
   BUILD_BUG_ON((op_flags) + 0U >= (1U << BIO_OP_SHIFT)); \
   ^
../fs/crypto/bio.c:134:3: note: in expansion of macro 'bio_set_op_attrs'
   bio_set_op_attrs(bio, REQ_OP_WRITE, REQ_NOENCRYPT);
   ^
  CC      crypto/asymmetric_keys/x509_cert_parser.o
make[3]: *** [../scripts/Makefile.build:339: fs/crypto/bio.o] Error 1
make[2]: *** [../scripts/Makefile.build:653: fs/crypto] Error 2
make[1]: *** [/home/lazerl0rd/recovery/kernel/sony/msm/Makefile:1105: fs] Error 2
make[1]: *** Waiting for unfinished jobs....
```

@bartcubbins seems to be correct about this part of the commit not applying to 4.9.

**Comment by lzlrd** [comment]
Building with ARM GCC 8.2.1 however, results in:

```
In file included from ../include/linux/linkage.h:4,
                 from ../include/linux/kernel.h:6,
                 from ../fs/ext4/readpage.c:30:
../fs/ext4/readpage.c: In function 'ext4_mpage_readpages':
../include/linux/compiler.h:530:38: error: call to '__compiletime_assert_301' declared with attribute error: BUILD_BUG_ON failed: (ctx ? (1ULL << __REQ_NOENCRYPT) : 0) + 0U >= (1U << BIO_OP_SHIFT)
  _compiletime_assert(condition, msg, __compiletime_assert_, __LINE__)
                                      ^
../include/linux/compiler.h:513:4: note: in definition of macro '__compiletime_assert'
    prefix ## suffix();    \
    ^~~~~~
../include/linux/compiler.h:530:2: note: in expansion of macro '_compiletime_assert'
  _compiletime_assert(condition, msg, __compiletime_assert_, __LINE__)
  ^~~~~~~~~~~~~~~~~~~
../include/linux/bug.h:54:37: note: in expansion of macro 'compiletime_assert'
 #define BUILD_BUG_ON_MSG(cond, msg) compiletime_assert(!(cond), msg)
                                     ^~~~~~~~~~~~~~~~~~
../include/linux/bug.h:78:2: note: in expansion of macro 'BUILD_BUG_ON_MSG'
  BUILD_BUG_ON_MSG(condition, "BUILD_BUG_ON failed: " #condition)
  ^~~~~~~~~~~~~~~~
../include/linux/blk_types.h:117:3: note: in expansion of macro 'BUILD_BUG_ON'
   BUILD_BUG_ON((op_flags) + 0U >= (1U << BIO_OP_SHIFT)); \
   ^~~~~~~~~~~~
../fs/ext4/readpage.c:300:4: note: in expansion of macro 'bio_set_op_attrs'
    bio_set_op_attrs(bio, REQ_OP_READ,
    ^~~~~~~~~~~~~~~~
make[3]: *** [../scripts/Makefile.build:339: fs/ext4/readpage.o] Error 1
make[3]: *** Waiting for unfinished jobs....
  CC      lib/vsprintf.o
  CC      lib/win_minmax.o
  CC      kernel/trace/trace_probe.o
  GEN     lib/crc32table.h
  LD      drivers/dma/qcom/built-in.o
  CC      drivers/crypto/msm/ota_crypto.o
  LD      net/ipc_router/built-in.o
  LD      drivers/dma/xilinx/built-in.o
  LD      drivers/dma/built-in.o
  CC      drivers/dma-buf/dma-buf.o
  CC      net/ipv4/route.o
make[2]: *** [../scripts/Makefile.build:653: fs/ext4] Error 2
make[1]: *** [/home/lazerl0rd/recovery/kernel/sony/msm/Makefile:1105: fs] Error 2
make[1]: *** Waiting for unfinished jobs....
```

**Comment by nathanchance** [comment]
It looks like `bio_set_op_attrs` was changed in 4.10: https://git.kernel.org/linus/ef295ecf090d3e86e5b742fc6ab34f1122a43773.

The `fs/crypto` change might not be needed, I guess only runtime will tell.

Whenever you backport something, it's important to understand _why_ you are getting a build failure when something changes because that can mean there are other changes that you aren't accounting for. Each Linux version sees 10,000 commits and hundreds of thousands of lines of code changed.

```
$ git diff --shortstat v4.9..v4.14
 40340 files changed, 4042290 insertions(+), 1346446 deletions(-)
```

**Comment by lzlrd** [comment]
> It looks like `bio_set_op_attrs` was changed in 4.10: https://git.kernel.org/linus/ef295ecf090d3e86e5b742fc6ab34f1122a43773.
> 
> The `fs/crypto` change might not be needed, I guess only runtime will tell.
> 
> Whenever you backport something, it's important to understand _why_ you are getting a build failure when something changes because that can mean there are other changes that you aren't accounting for. Each Linux version sees 10,000 commits and hundreds of thousands of lines of code changed.
> 
> ```
> $ git diff --shortstat v4.9..v4.14
>  40340 files changed, 4042290 insertions(+), 1346446 deletions(-)
> ```

Pavel guided me and I've seen that commit, I just thought that you did too and maybe there was an exception/we needed to change something here. I'm happy you brought it up tho.

File: fs/ext4/readpage.c
Comment (bartcubbins) [inline fs/ext4/readpage.c]: @lazerl0rd leave the old parameter. then pr ready for merge

**Review by bartcubbins — CHANGES_REQUESTED** [review]

File: fs/f2fs/data.c
Comment (lzlrd) [inline fs/f2fs/data.c]: I'll need to change this too @bartcubbins.

File: fs/ext4/readpage.c
Comment (lzlrd) [reply fs/ext4/readpage.c]: Done

File: fs/f2fs/data.c
Comment (lzlrd) [reply fs/f2fs/data.c]: Done

**Comment by pablomh** [comment]
Hi @lazerl0rd, can you please rebase?

**Comment by lzlrd** [comment]
@pablomh just as soon as I build and test it, I will :).

**Comment by lzlrd** [comment]
@pablomh @MartinX3 @jerpelea I rebased the commits.

**Review by pablomh — COMMENTED** [review]
Fails to build:

Already defined at line 192:

```
In file included from /repos/android/SonyAosp/android-9.0.0/kernel/sony/msm-4.9/include/linux/fs.h:33:
/repos/android/SonyAosp/android-9.0.0/kernel/sony/msm-4.9/include/linux/blk_types.h:222:2: error: redefinition of enumerator '__REQ_NOENCRYPT'
        __REQ_NOENCRYPT,        /* ok to not encrypt (already encrypted at fs level) */
        ^
/repos/android/SonyAosp/android-9.0.0/kernel/sony/msm-4.9/include/linux/blk_types.h:192:2: note: previous definition is here
        __REQ_NOENCRYPT,        /* ok to not encrypt (already encrypted at fs
        ^
1 error generated.
make[2]: *** [/repos/android/SonyAosp/android-9.0.0/kernel/sony/msm-4.9/./Kbuild:57: arch/arm64/kernel/asm-offsets.s] Error 1
make[1]: *** [/repos/android/SonyAosp/android-9.0.0/kernel/sony/msm-4.9/Makefile:1140: prepare0] Error 2
make[1]: Leaving directory '/repos/android/SonyAosp/android-9.0.0/out/target/product/kugo/obj/KERNEL_OBJ'
make: *** [Makefile:152: sub-make] Error 2
make: Leaving directory '/repos/android/SonyAosp/android-9.0.0/kernel/sony/msm-4.9'
```

**Review by pablomh — COMMENTED** [review]
Unneeded commit (seems to be already in place).

**Comment by pablomh** [comment]
It seems to me that most of these patches are the same (in spirit) that the ones already merged in https://github.com/sonyxperiadev/kernel/pull/1962.

Can you please double-check which ones are really needed and explain which functionality is missing and fixed by them?

**Comment by lzlrd** [comment]
@pablomh those commits in #1962 got force-removed.

I made a mistake that killed build for the entire kernel, which had to be removed ASAP.

**Comment by lzlrd** [comment]
However, you are right @pablomh that https://github.com/sonyxperiadev/kernel/commit/0c4faf89d2ff992d79543475f9652d22103a77e9 and https://github.com/sonyxperiadev/kernel/commit/5ae2d575236231694553eb848f6081798d2b9107 were already existing commits. I'll clean-up this PR for [hopefully the final time] and get it ready for a merge.

**Comment by lzlrd** [comment]
@pablomh it's ready :P.

**Comment by pablomh** [comment]
I'm still able to use adaptable storage in my pioneer (nile) with this PR applied, so it's good from my point of view.

**Comment by kholk** [comment]
I've checked the code against k4.14 and this seems to be correctly ported,
but I cannot 100% guarantee that all the required pieces are there on k4.9.

Please test this on a UFS storage equipped device (any Yoshino or any Tama):
if the testing goes well, then from my perspective it can be merged.

**Comment by MartinX3** [comment]
@kholk what exactly should I test after compiling this PR?
And could my userdata get lost?

**Comment by lzlrd** [comment]
@MartinX3 I highly doubt your userdata will be lost (maybe not decrypted, but at one point this PR could corrupt data so I’d say to backup anyways). Test SDCard usage and if in Storage Settings there’s an option to convert/migrate/format as internal/format as adoptable setting under SDCard. By SDCard I mean the physical thing not /sdcard.

Also test file access on /data and your apps to make sure nothing random is broken. Sorry this is vague but I would have performed an I/O bench (checking regressions), hashing test on files before and after (randomly chosen files that won’t change in /data and /data/media/0), and encryption on and off (ICE FBE).

**Comment by kholk** [comment]
Yeah, it may happen, so of course backup your data.... but it's very unlikely, I
wouldn't be worried about that.

I'd be happy if you can test this so that we will be able to finally get this thing
merged: we're not too far from the feature freeze and we have to sort this
out before the dev window gets closed.

**Comment by MartinX3** [comment]
@kholk can it get tested together with your vmalloc PR or should I wait until it gets merged?
And good to know about the feature freeze window.

**Comment by kholk** [comment]
Yes it can!

**Comment by pablomh** [comment]
Please remove the last commit from the PR.

They'd need a new PR and discussion.

**Comment by lazerl0rd** [comment]
@pablomh if you don't mind corruption and no booting, sure. It has been discussed here with @nathanchance and @bartcubbins, look under the hidden/resolved comments.

**Comment by lazerl0rd** [comment]
Oops @pablomh my bad thanks.

**Comment by lzlrd** [comment]
> Please remove the last commit from the PR.
> 
> They'd need a new PR and discussion.

@pablomh apologies, I pushed to the wrong branch. It has been removed.

**Comment by lzlrd** [comment]
> Yes it can!

@jerpelea could this get merged? I've tested it for weeks now, and no one seems to have anything against it.

**Pull request merged**
